lib/rust/ensogl/core/src/debug/stats.rs

```diff
@@ -15,48 +15,137 @@
 
 use enso_prelude::*;
 
+use enso_web::Performance;
 use js_sys::ArrayBuffer;
 use js_sys::WebAssembly::Memory;
 use wasm_bindgen::JsCast;
 
 
 
+// ====================
+// === TimeProvider ===
+// ====================
+
+/// An entity that can retrieve current time.
+pub trait TimeProvider {
+    /// Returns current time, measured in milliseconds.
+    fn now(&self) -> f64;
+}
+
+impl TimeProvider for Performance {
+    fn now(&self) -> f64 {
+        self.now()
+    }
+}
+
+
+
 // =============
 // === Stats ===
 // =============
 
-/// Structure containing all the gathered stats.
-#[derive(Debug, Clone, CloneRef)]
-pub struct Stats {
-    rc: Rc<RefCell<StatsData>>,
+/// Tracks stats related to the current rendering frame, providing methods for modifying and
+/// retrieving their values.
+pub type Stats = StatsOverTime<Performance>;
+
+
+// === StatsOverTime ===
+
+/// Tracks stats related to the current rendering frame, providing methods for modifying and
+/// retrieving their values.
+#[derive(Debug, CloneRef)]
+pub struct StatsOverTime<T: TimeProvider> {
```

Comment:
Please do not use constraints in structs definition. I was explaining it multiple times in different PRs. Such constraints leak everywhere. For example, you implement later `Clone` for this structure. Is it really necessary that `T:TimeProvider` when cloning the structure?

There are big topics about it on Haskell mailing lists and it is considered a big anti-pattern in Haskell (which is supposed to be banned). I believe that we should ban it too.